src/tracing/tracetool/parsetool.py

```diff
@@ -0,0 +1,472 @@
+                        # Make non-basic types strings and implement a method operator std::string()
+                        out(fobj, '       ctf_string(' + n + ', ' + n + ')')
+
+                out(fobj, '   )',
+                    ')',)
+
+            else:
+                out(fobj, 'TRACEPOINT_EVENT(',
+                    '   %(subsys)s,',
+                    '   %(name)s,',
+                    '   TP_ARGS(void),',
+                    '   TP_FIELDS()',
+                    ')',
+                    subsys=e.subsys,
+                    name=e.name)
+
+            out(fobj, 'TRACEPOINT_LOGLEVEL(%(subsys)s, %(name)s, %(loglevel)s)',
+                '',
+                subsys=e.subsys,
+                name=e.name,
+                loglevel=LTTNG_LOGLEVEL_MAPPING[0 if e.level < 0 else min(int(e.level), len(LTTNG_LOGLEVEL_MAPPING) - 1)])
+
+        for v in fhandles.values():
+            v.close()
+
+
+    def generate_c(self, events, outdir):
+        fhandles = {}
+        for e in events:
+            if e.subsys not in fhandles:
+                fhandles[e.subsys] = open(outdir + "/tracing/" + e.subsys + ".c", "w")
+                out(fhandles[e.subsys],
+                    '/* This file was generated automatically by the tracetool script.',
+                    '   Do not edit it manually. Refer to the documentation to add logging. */',
+                    '')
+                out(fhandles[e.subsys],
+                    '#define TRACEPOINT_CREATE_PROBES',
+                    ''
+                    '/*',
+                    ' * The header containing our TRACEPOINT_EVENTs.',
+                    ' */',
+                    '#include "tracing/%(subsys)s.h"',
+                    subsys=e.subsys)
+
+        for v in fhandles.values():
+            v.close()
+
+
+    def generate_header(self, events, outdir):
+        fhandles = {}
+        for e in events:
+            if e.subsys not in fhandles:
+                fhandles[e.subsys] = open(outdir + "/include/tracing/" + e.subsys + "_impl.h", "w")
+                fobj = fhandles[e.subsys]
+                out(fobj,
+                    '/* This file was generated automatically by the tracetool script.',
+                    '   Do not edit it manually. Refer to the documentation to add logging. */',
+                    '')
+                out(fobj,
+                    '#ifndef %(usubsys)s_IMPL',
+                    '#define %(usubsys)s_IMPL',
+                    '',
+                    '#include "tracing/%(subsys)s.h"',
+                    usubsys=e.subsys.upper(),
+                    subsys=e.subsys)
+
+            fobj = fhandles[e.subsys]
+            # Write the event's function
+            out(fobj, '',
+                'static inline void __%(api)s(%(args)s)',
+                '{',
+                api=e.api(e.CEPH_TRACE),
+                args=", ".join([' '.join(arg[:2]) if is_valid_type(arg[0]) else ' '.join(('string', arg[1])) for arg in e.args]))
```

Comment:
might want to differentiate the args which are already `std::string` or `string` in `Event.build()`. probably we could have three `Arg` types to represent them?

so we don't create a temporary copy of them just for passing them to `ctf_string()`.